Data processing code for ALARA flux input

anu1217 opened the pull request:
Wrote a Python script that takes OpenMC neutron fluxes and converts them into a form suitable for ALARA. These fluxes are originally from the spherical shell problem.

anu1217 committed:
Data processing code for ALARA flux input

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+
+#Loading the csv that contains the energy bounds of the Vitamin J structure 
+Vit_J = 'VitaminJEnergyGroupStructure.csv'
+Vit_J_Groups = pd.read_csv(Vit_J, header=1)
```
Since this is pretty constant, we probably don't need the intermediate variable
```suggestion
Vit_J_Groups = pd.read_csv('VitaminJEnergyGroupStructure.csv', header=1)
```

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+#Loading the csv file that contains the energy bins and neutron flux values (from OpenMC)
+#This csv file was created in a Python code that reads the h5 output from OpenMC
+Flux = 'Neutron_Flux.csv'
+Flux_Data = pd.read_csv(Flux, header=1)
```
How did you get this data out of OpenMC?  Probably another python script. Maybe we should put those things together?

anu1217 replied:
Added this to the script.

gonuke commented on VitaminJEnergyGroupStructure.csv:
```diff
@@ -0,0 +1,178 @@
+﻿Bin #,Energy Bound [eV]
+,
```
Why this empty row?

anu1217 replied:
Removed empty row in Python script.

gonuke commented on VitaminJEnergyGroupStructure.csv:
Do we need the bin number?  Could this just be a text file with the energy bounds instead of also including the group numbers?

anu1217 replied:
I think having bin numbers would make it easier to refer to an energy bound even when the file is not read in Python.

gonuke replied:
Ok

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+
+for i in range(len(x) - 1):
+    lower_bound = x[i]
+    upper_bound = x[i + 1]
```
```suggestion
for lower_bound, upper_bound in zip(x[:-1], x[1:]):
```

anu1217 replied:
This loop was taken out

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+#Loading the csv that contains the energy bounds of the Vitamin J structure 
+Vit_J = 'VitaminJEnergyGroupStructure.csv'
+Vit_J_Groups = pd.read_csv(Vit_J, header=1)
+x_lh = Vit_J_Groups.iloc[:, 1]
```
Use more descriptive variable names, like `ebounds`

anu1217 replied:
Variable names have been updated

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+Vit_J_Groups = pd.read_csv(Vit_J, header=1)
+x_lh = Vit_J_Groups.iloc[:, 1]
+
+x = sorted(x_lh, reverse=True) #Energy bounds arranged from high energy to low energy
```
Use more descriptive variable name like `ebounds`

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+#the Vitamin J structure
+
+#Initializing the array of fluxes that are summed up
+flux_sums = [0] * (len(x) - 1)
```
Use `np.zeros`

anu1217 replied:
Initialization changed to np.zeros(175)

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,79 @@
+#Initializing the array of fluxes that are summed up
+flux_sums = [0] * (len(x) - 1)
+
+#Iterating over each interval of energies in the Vit J structure
```
Notwithstanding the fact that you can avoid the whole thing by making OpenMC group structure match Vitamin-J....

This nested loop approach is a little cumbersome.  I'd recommend that you do something like:
```
bin = 0

for result_ebound, result_flux in results:
    if result_ebound > ebounds[bin]:
        bin +=1
    flux[bin] += result_flux
```

anu1217 replied:
Approached this using a for-while-if statement

gonuke reviewed (changes requested):
A few suggestions of varying kinds

anu1217 committed:
Update VitaminJProcessing.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

anu1217 committed:
Update VitaminJEnergyGroupStructure.csv

anu1217 committed:
Update VitaminJProcessing.py

anu1217 committed:
Update VitaminJProcessing.py

anu1217 committed:
Update VitaminJProcessing.py

gonuke commented:
I recommend slightly more descriptive commit messages - they don't have to be long, but should be more descriptive of what's happening

anu1217 committed:
Removed header entry in Vit_J and Flux_Data

anu1217 committed:
Replaced nested for loops with single for loop

- Initialized flux_sums with np.zeros()

- Added a for-while-if statement to represent the flux summation/iteration over the bin number

anu1217 committed:
Added code to generate Neutron_Flux.csv file

anu1217 committed:
Added clarifying statement about where OpenMC is used

anu1217 committed:
Readded conversion from flux list to array format

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+import openmc
+
+#Obtaining results from statepoint file:
+with openmc.StatePoint('statepoint.10.h5') as sp:
+    fl = sp.get_tally(name="Flux spectrum")
```
It's certainly possible that this filename may vary by use case.  Is it also possible that the tally name might vary?

Perhaps these should be command line variables using `argparse`, with default values as shown here.

gonuke replied:
recommend longer variable names for readability: `fux_tally` instead of `fl`

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+with openmc.StatePoint('statepoint.10.h5') as sp:
+    fl = sp.get_tally(name="Flux spectrum")
+    # Get the neutron energies from the energy filter
+    energy_filter_fl = fl.filters[0]
+    energies_fl = energy_filter_fl.bins[:, 0]
```
Could probably combine this into one line.  Are these upper bounds or lower bounds (I assume upper).
```suggestion
    upper_ebounds = flux_tally.filters[0].bins[:, 0]
```

gonuke commented on Neutron_Flux.csv:
```diff
@@ -0,0 +1,502 @@
+Neutron Energy [eV],Flux [#-cm/sp]
+,
```
where does this empty line come from?

gonuke commented on Neutron_Flux.csv:
Is this file output?  If so, it might not belong in the repo unless it's part of an example of how to use it or part of a unit test.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+# Save neutron flux vs energy as csv file:
+np.savetxt('Neutron_Flux.csv', np.c_[energies_fl, flux],delimiter=',')
+
+#The rest of this code does not require OpenMC
```
This comment may not be important.  Alternatively, you could have a function to do the OpenMC parsing and only import OpenMC within the scope of that function.  In general, I'm not sure that it matters whether the use of an imported module is confined to some scope, however.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+results = zip(energy, flux)
+
+#Iterating over each interval of energies in the Vit J structure
+#Since there are 175 energy bins, there are 175 - 1 = 174 intervals
```
This is not true.  Each bin is an interval - there are 175 bins/intervals.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+#the Vitamin J structure
+
+#Initializing the array of fluxes that are summed up
+flux_sums = np.zeros(175)
```
magic number.  Either define what 175 is in a variable, or use the length of `Vit_J`.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+
+ebounds = sorted(ebounds_lh, reverse=True) #Energy bounds arranged from high energy to low energy
+        
+#Loading the csv file that contains the energy bins and neutron flux values (from OpenMC)
+#This csv file was created in a Python code that reads the h5 output from OpenMC
+Flux_Data = pd.read_csv('Neutron_Flux.csv')
+# Extract both columns from the CSV file
+energy_lh = Flux_Data.iloc[:, 0]
+energy = sorted(energy_lh, reverse=True) #Energy bins arranged from highest to lowest energy
+flux_lh = Flux_Data.iloc[:, 1]
+flux = sorted(flux_lh, reverse=True) #Fluxes arranged to highest to lowest corresponding energy
```
Why are you reading this in again from the CSV - you already have it in memory above.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+groups = len(ebounds) - 1
+bin = 0
+
+for energy, flux in results:
+        while bin < groups and ebounds[bin + 1] >= energy:
+                bin += 1
+        if bin < len(flux_sums):
+                flux_sums[bin] += flux
```
A: this whole approach should be unnecessary if you define your flux bins in OpenMC to match Vitamin-J - this is unequivocally the correct way to do this!  Nevertheless, I will provide comments about logic/readability.
B: Indentation doesn't seem quite right

Detailed comments:
The logic here is not very readable.
* you have the same information represented in different ways: `groups` and `len(flux_sums)` are definitely related and perhaps should only be used in one way
* it looks like you are treating `ebound` as a lower bound, but it may be an upper bound? 
* as you loop through the results, you should only ever change the coarse bin boundary by one bin at a time and only occasionally - although a `while` loop may result in correct behavior, it implies a far more complicated logic than is truly necessary
* If you are concerned about whether the data in `results` is outside the energy range of your Vitamin-J coarse structure, you can check this before the loop and only include data that is within the right range, rather than testing at each step
* This algorithm does not correctly handle situations where the fine bins straddle the coarse bins and the fine results needs to be distributed between the two coarse bins (more reason to run OpenMC with the final bin structure)

anu1217 replied:
I'll add the Vit-J structure to the OpenMC input. We can discuss some of the feedback about this code

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+
+# Saving flux_sums as a 30x6 tab-delimited file:
+    
+Array_1 = flux_sums_array[:174] #The first 174 elements of the flux summation array
```
Magic numbers: 174 (which should probably be 175, see above) should be defined as a variable and/or related to some other variable.  You already have `groups` which is probably equal to this (although perhaps incorrectly)

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+# Saving flux_sums as a 30x6 tab-delimited file:
+    
+Array_1 = flux_sums_array[:174] #The first 174 elements of the flux summation array
+print((Array_1))
```
`print` statements may be handy during development, but should rarely be in code merged into the final product.  If you really think it's important to have output, use the `logging` module that makes it easy to suppress it.

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+flux_sums_shaped = Array_1.reshape(29,6) #Accounts for the first 174 elements
+
+# Open the file in append mode and write each row of the reshaped array
+with open('Flux_Summation.txt', 'w') as f:
```
This filename may be a user option: `argparse`

gonuke commented on VitaminJProcessing.py:
```diff
@@ -0,0 +1,88 @@
+# Converting list of fluxes to array format    
+flux_sums_array = np.array(flux_sums)
+
+# Saving flux_sums as a 30x6 tab-delimited file:
```
First, ALARA doesn't actually require this (I'm pretty sure) but I agree it makes it a little more readable.

Second, there may be an easier way to do this:
```
vals_per_row = 6
for first_val_in_row in range(0, len(flux_sums_array), vals_per_row):
    formatted_row = ' '.join([f'{value:.5e}' for value in flux_sums_array[first_val_in_row:first_val_in_row+vals_per_row]])
    f.write(formatted_row)
```

gonuke reviewed (changes requested):
There are many comments here based on our CNERG expectations for readability and reusability, two things we value a great deal since a lot of code gets handed from one person to another.

Here are some high level concepts to consider (and discuss if you have questions):
* modularity & separation of concerns: code should be divided into separate methods/functions each with a clear purpose.  Function names are often verbs that clearly define what they do to variables (that are often nouns).  With clear function names that represent clear and narrow purpose (separation of concerns), it is can be obvious what a piece of code is doing.  
* (related) nearly all code should be within a method/function and not at global scope - this facilitates testing and reuse of that capability. Any code at global scope should have a clear reason that it must be there
* magic number: almost all numbers should be defined with a variable name for two reasons: (1) if they ever need to change the value you only need to change it in one place and ensure continued correctness/consistency, (2) it makes it more readable and communicates the purpose of the number to everyone
* `argparse` should be part of all new projects to make it immediately possible to change some values (e.s. especially file names) from the command line
